sdk/core/azure-core/azure/core/pipeline/_tools_async.py

```diff
@@ -39,10 +53,13 @@ async def await_result(func, *args, **kwargs):
     :rtype: any
     :return: The result of the function
     """
-    result = func(*args, **kwargs)
+    result: Union[T, Awaitable[T]] = func(*args, **kwargs)
+    # pyright has issue with narrowing types here
+    # https://github.com/microsoft/pyright/issues/5860
     if hasattr(result, "__await__"):
-        # type ignore on await: https://github.com/python/mypy/issues/7587
-        return await result  # type: ignore
+        result = cast(Awaitable[T], result)
```

Comment:
Seems silly that we have to cast here. 
Maybe we wouldn't have to if we used `isinstance(result, Awaitable)`?
I think the subclass check for `abc.Awaitable` is basically only checking for the `__await__` method anyway.

Reply:
Yes, that works, I initially didn't want to change the implementation, but works for me :)